src/session/chat/action_list.rs

```diff
@@ -0,0 +1,147 @@
+    }
+
+    impl ObjectImpl for ChatActionList {
+        fn properties() -> &'static [glib::ParamSpec] {
+            static PROPERTIES: Lazy<Vec<glib::ParamSpec>> = Lazy::new(|| {
+                vec![glib::ParamSpecObject::new(
+                    "chat",
+                    "Chat",
+                    "The chat relative to this chat action list",
+                    Chat::static_type(),
+                    glib::ParamFlags::READWRITE | glib::ParamFlags::CONSTRUCT_ONLY,
+                )]
+            });
+            PROPERTIES.as_ref()
+        }
+
+        fn set_property(
+            &self,
+            _obj: &Self::Type,
+            _id: usize,
+            value: &glib::Value,
+            pspec: &glib::ParamSpec,
+        ) {
+            match pspec.name() {
+                "chat" => self.chat.set(Some(&value.get().unwrap())),
+                _ => unimplemented!(),
+            }
+        }
+
+        fn property(&self, obj: &Self::Type, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+            match pspec.name() {
+                "chat" => obj.chat().to_value(),
+                _ => unimplemented!(),
+            }
+        }
+    }
+
+    impl ListModelImpl for ChatActionList {
+        fn item_type(&self, _list_model: &Self::Type) -> glib::Type {
+            ChatAction::static_type()
+        }
+
+        fn n_items(&self, _list_model: &Self::Type) -> u32 {
+            self.list.borrow().len() as u32
+        }
+
+        fn item(&self, _list_model: &Self::Type, position: u32) -> Option<glib::Object> {
+            self.list
+                .borrow()
+                .get_index(position as usize)
+                .map(|(_, obj)| obj.upcast_ref())
+                .cloned()
+        }
+    }
+}
+
+glib::wrapper! {
+    pub struct ChatActionList(ObjectSubclass<imp::ChatActionList>)
+        @implements gio::ListModel;
+}
+
+impl From<&Chat> for ChatActionList {
+    fn from(chat: &Chat) -> Self {
+        glib::Object::new(&[("chat", chat)]).expect("Failed to create ChatActionList")
+    }
+}
+
+impl ChatActionList {
+    pub fn handle_update(&self, update: types::UpdateChatAction) {
+        let imp = self.imp();
+
+        let sender_id = match &update.sender_id {
+            enums::MessageSender::User(user) => user.user_id,
+            enums::MessageSender::Chat(chat) => chat.chat_id,
+        };
+
+        if let Some((position, ..)) = imp.list.borrow_mut().shift_remove_full(&sender_id) {
+            self.items_changed(position as u32, 1, 0);
+        }
+
+        match update.action {
+            enums::ChatAction::Cancel => {}
+            action => {
+                imp.list.borrow_mut().insert(
+                    sender_id,
+                    ChatAction::new(action, &update.sender_id, &self.chat()),
+                );
+
+                self.items_changed(self.n_items() - 1, 0, 1);
+            }
+        }
+
+        self.chat().notify("actions");
```

Comment:
I don't really like having this for two reasons: we notify for a property of a parent object and we notify the property even if it's not changing, but the data inside it is. I think we should use the items-changed signal for updating the widget instead of just the notify. Actually, having a [SignalExpression](https://gitlab.gnome.org/GNOME/gtk/-/issues/4709) would probably be the best thing, but that doesn't currently exist.

Anyway, for now I think this is fine. I'd just suggest adding a TODO comment that says that this should probably be removed in the future for the reasons I said above.

Reply:
I solved the first problem by now triggering the notification directly in the chat. 

I approached the matter somewhat pragmatically: I think, mathematically, the value of the variable "actions" also changes when elements in the Set change. So I just used that as an excuse to be able to use PropertyExpressions. :-P

But you are right, of course: it is more efficient in certain cases to handle modifications of the list selectively. I wrote that again as a comment in `row.rs:213`.